python/ray/serve/router.py

```diff
@@ -41,382 +43,190 @@ class Query:
-            updated_replica_tags = set(replica_dict.keys())
-            curr_replica_tags = {
-                tag.replace(backend_tag + ":", "")
-                for tag in self.replicas.keys() if tag.startswith(backend_tag)
-            }
-
-            added_replicas = updated_replica_tags - curr_replica_tags
-            removed_replicas = curr_replica_tags - updated_replica_tags
-
-            for replica_tag in added_replicas:
-                await self.add_new_replica(backend_tag, replica_tag,
-                                           replica_dict[replica_tag])
-            for replica_tag in removed_replicas:
-                await self.remove_replica(backend_tag, replica_tag)
-
-    async def update_backend_configs(self, backend_configs):
-        updated_backends = set(backend_configs.keys())
-        curr_backends = set(self.backend_info.keys())
-
-        for backend in updated_backends:
-            await self.set_backend_config(backend, backend_configs[backend])
-
-        removed_backends = curr_backends - updated_backends
-        for backend in removed_backends:
-            await self.remove_backend(backend)
-
-    async def enqueue_request(self, request_meta, *request_args,
-                              **request_kwargs):
-        endpoint = request_meta.endpoint
-        logger.debug("Received request {} for endpoint {}.".format(
-            request_meta.request_id, endpoint))
-        request_start = time.time()
-        self.num_router_requests.record(1, tags={"endpoint": endpoint})
 
-        request_context = request_meta.request_context
+    async def setup_in_async_loop(self):
+        # NOTE(simon): we separated this method to run all the setup in asyncio
+        # loop.
+        self.long_pull_client = LongPollerAsyncClient(
+            self.controller, {
+                "traffic_policies": self._update_traffic_policies,
+                "worker_handles": self._update_worker_handles,
+                "backend_configs": self._update_backend_configs,
+            })
+
+    async def _update_traffic_policies(self, traffic_policies):
+        for endpoint, traffic_policy in traffic_policies.items():
+            self.endpoint_policies[endpoint] = RandomEndpointPolicy(
+                traffic_policy)
+            if endpoint in self._pending_endpoints:
+                event = self._pending_endpoints.pop(endpoint)
+                event.set()
+
+    async def _update_worker_handles(self, worker_handles):
+        for backend_tag, replica_handles in worker_handles.items():
+            self.backend_replicas[backend_tag].update_worker_replicas(
+                replica_handles)
+
+    async def _update_backend_configs(self, backend_configs):
+        for backend_tag, config in backend_configs.items():
+            self.backend_replicas[backend_tag].set_max_concurrent_queries(
+                config.max_concurrent_queries)
+
+    async def enqueue_request(
```

Comment:
there aren't really queues anymore, maybe we should rename this :) --> `assign_request`? `send_request`?